web/src/lib/components/memory-page/memory-viewer.svelte

```diff
@@ -42,84 +83,51 @@
 
 	$: canAdvance = !!(nextMemory || nextAsset);
 
-	$: if (!canAdvance && browser) {
-		pause();
-	}
+	const toNextMemory = () => goto(`?memory=${memoryIndex + 1}`);
+	const toPreviousMemory = () => goto(`?memory=${memoryIndex - 1}`);
 
-	let memoryGallery: HTMLElement;
-	let memoryWrapper: HTMLElement;
-	let galleryInView = false;
+	const toNextAsset = () => goto(`?memory=${memoryIndex}&asset=${assetIndex + 1}`);
+	const toPreviousAsset = () => goto(`?memory=${memoryIndex}&asset=${assetIndex - 1}`);
 
-	onMount(async () => {
-		if (!$memoryStore) {
-			const { data } = await api.assetApi.getMemoryLane({
-				timestamp: DateTime.local().startOf('day').toISO()
-			});
-			$memoryStore = data;
-		}
-	});
-
-	onDestroy(() => browser && pause());
-
-	const toPreviousMemory = () => previousMemory && goto(`?memory=${memoryIndex - 1}`);
+	const toNext = () => (nextAsset ? toNextAsset() : toNextMemory());
+	const toPrevious = () => (previousAsset ? toPreviousAsset() : toPreviousMemory());
 
-	const toNextMemory = () => nextMemory && goto(`?memory=${memoryIndex + 1}`);
+	const animation = new Animation(5000); // 5 second duration
 
-	const toPreviousAsset = () =>
-		previousAsset ? goto(`?memory=${memoryIndex}&asset=${assetIndex - 1}`) : toPreviousMemory();
+	let { progress } = animation;
 
-	const toNextAsset = () =>
-		nextAsset ? goto(`?memory=${memoryIndex}&asset=${assetIndex + 1}`) : toNextMemory();
-
-	const duration = 5000; // 5 seconds
+	$: if ($progress === 1) {
+		toNext();
+	}
 
 	let paused = true;
-	let progress = 0;
-	let animationFrameRequest: number;
-	let start: number | null = null;
-
-	const requestDraw = () => (animationFrameRequest = requestAnimationFrame(draw));
 
-	const draw = (now: number) => {
-		requestDraw();
+	$: paused ||= !canAdvance;
 
-		start ??= now - progress * duration;
-
-		const elapsed = now - start;
-		progress = Math.min(1, elapsed / duration);
+	$: if (paused) {
+		animation.cancel();
+		animation.resetStart();
+	} else {
+		animation.request();
+	}
 
-		if (progress !== 1) {
-			return;
-		}
+	// Progress should be reset when the current memory or asset changes.
+	$: memoryIndex, assetIndex, animation.reset();
```

Comment:
Interesting syntax, I have not seen this before, what does it do?

Reply:
It listens for changes to `memoryIndex` and `assetIndex` then calls `animation.reset()` if either of them change.

https://twitter.com/liyuanqiu/status/1149235193296773122